test/hotspot/jtreg/serviceability/HeapDump/IntegrityHeapDumpTest.java

```diff
@@ -81,7 +81,7 @@ private static void attachDumpAndVerify(File heapDumpFile,
                 .addToolArg("-parallel=" + Runtime.getRuntime().availableProcessors())
                 .addToolArg(heapDumpFile.getAbsolutePath());
 
-        ProcessBuilder processBuilder = SATestUtils.createProcessBuilder(launcher);
+        ProcessBuilder processBuilder = new ProcessBuilder(launcher.getCommand());
```

Comment:
Can you explain this change? This will result in the test not being run with sudo in cases where sudo is necessary. Before we get here we already checked if we need sudo (not running as root) and verified that passwordless sudo is supported. Otherwise the test is skipped and would never get here. So if sudo is required, the test will fail to attach to the debuggee. Note you might not see this issue if you have DevSecurityMode enabled. See [JDK-8313357](https://bugs.openjdk.org/browse/JDK-8313357)

Reply:
I see now that this is a new test, and not an SA test, so probably should never have been using SATestUtils in the first place (I'm curious as to how that ended up happening). You should also remove the import of SATestUtils.

Reply:
> Can you explain this change? This will result in the test not being run with sudo in cases where sudo is necessary. Before we get here we already checked if we need sudo (not running as root) and verified that passwordless sudo is supported. Otherwise the test is skipped and would never get here. So if sudo is required, the test will fail to attach to the debuggee. Note you might not see this issue if you have DevSecurityMode enabled. See [JDK-8313357](https://bugs.openjdk.org/browse/JDK-8313357)

I followed what DuplicateArrayClassesTest.java. The reason seems that `SATestUtils.createProcessBuilder` does not touch any verification of whether passwordless sudo is supported:

```java
public class SATestUtils {
    public static ProcessBuilder createProcessBuilder(JDKToolLauncher launcher) {
        List<String> cmdStringList = Arrays.asList(launcher.getCommand());
        if (needsPrivileges()) {
            cmdStringList = addPrivileges(cmdStringList);
        }
        return new ProcessBuilder(cmdStringList);
    }
    public static boolean needsPrivileges() {
        return Platform.isOSX() && !Platform.isRoot();
    }
    private static List<String> addPrivileges(List<String> cmdStringList) {
        if (!Platform.isOSX()) {
            throw new RuntimeException("Can only add privileges on OSX.");
        }

        System.out.println("Adding 'sudo -E -n' to the command.");
        List<String> outStringList = new ArrayList<String>();
        outStringList.add("sudo");
        outStringList.add("-E"); // Preserve existing environment variables.
        outStringList.add("-n"); // non-interactive. Don't prompt for password. Must be cached or not required.
        outStringList.addAll(cmdStringList);
        return outStringList;
    }
```

Reply:
But DuplicateArrayClassesTest.java does not use SATestUtils.createProcessBuilder(), so I'm still wondering how this use of SATestUtils.createProcessBuilder() got in this new test in the first place.

Yes, a test must call SATestUtils.skipIfCannotAttach() before calling SATestUtils.createProcessBuilder(). But  these calls should only be made by SA tests. Non SA tests have no need for them.

And just to clarify, there are two types of "attaches" going on here. SA attach does a debugger attach, which varies depending on the OS, but requires some sort of root or sudo priveledges on OSX.  jcmds and tools like jstack and jmap use the java Attach API, which is just a socket based connection with the JVM, and just requires that the attaching process be the same user as the JDK process.

Reply:
>  so I'm still wondering how this use of SATestUtils.createProcessBuilder() got in this new test in the first place

I simply grep existing tests and found this snippet to create java process. There is no special intention, it is an obvious misuse :-)